Accept terms of use

gahermar opened the pull request:
Accept term of use after login for all users
After change term of use set status accepted on all users to false.

MartinGaher committed:
Test task - 317 - Accept Terms of use for users

MartinGaher committed:
repair tests

MartinGaher committed:
tests

MartinGaher committed:
add headline + remove decline button

jstrojil commented on app/assets/stylesheets/application.css.scss:
```diff
@@ -221,6 +221,16 @@ form.schedule {
+
+.term-of-use-button {
+  margin: 10px;
+  float:right;
```
`float: right;`

jstrojil commented on app/controllers/application_controller.rb:
```diff
@@ -6,6 +6,7 @@ class ApplicationController < ActionController::Base
 
   before_action :load_site_configuration
   before_action :logout_inactive
+  before_action :terms_of_use
```
Does this need to run before every action in this controller?

gahermar replied:
My bad. I try to do it like in self registration. But forgot to look on other variables in application.html.haml. So i edit it and use existing @site there and remove before_action and terms_of_use function from application controller.

jstrojil commented on app/controllers/sites_controller.rb:
```diff
@@ -18,7 +18,14 @@ def show
   def update
-    if @site.update(site_params)
+    @site.attributes = site_params
+    if(@site.terms_of_use_content_changed?)
```
```
if @site.terms_of_use_content_changed?
```
We usually omit the unnecessary `()`

jstrojil commented on app/controllers/sites_controller.rb:
```diff
@@ -18,7 +18,14 @@ def show
+    @site.attributes = site_params
+    if(@site.terms_of_use_content_changed?)
+      id = current_user.id
+      users = User.where("id != ?", id)
```
```User.where.not(id: id)```

jstrojil commented on app/controllers/sites_controller.rb:
```diff
@@ -18,7 +18,14 @@ def show
+    if(@site.terms_of_use_content_changed?)
+      id = current_user.id
+      users = User.where("id != ?", id)
+      users.update_all(terms_of_use_accepted: false)
```
This is one possible way of course. What would you think about a solution that uses `terms_accepted_at (DateTime)` on a user model - this way you would simply compare `updated_at` on Site with the `terms_accepted_at` on a model to check whether the user has accepted the updated terms.

gahermar replied:
Yeah this solution is nice to. I following instructions in task (https://github.com/panterch/future_kids/issues/317)
And it depend ... if we need to save date of accepted it will be better your solution. but if we wouldn't need it, i think solution with boolean is better .... for example when we need check in DB who accept and who not it will be more readeble when we have this data. Of course we can have both or it is just simple sql query to show it but if we no need date of accept i think it is useles.

panterch replied:
Since you introduced terms_of_use_accepted_date the boolean field terms_of_use_accepted is a duplicate. I'd would stay with terms_of_use_accepted_date on the user and simply add a virtual field on user:

def terms_of_use_accepted
  user.terms_of_use_accepted_date < site.terms_of_use_accepted_date 
end


Then you don't have to update all database rows when the terms change.

gahermar replied:
yes, good tip. I changed it..

jstrojil commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,19 @@
+class UsersController < ApplicationController
+
+  def edit_terms
+    if(current_user != nil)
```
This project is using `cancancan` and `devise` - I think there are better ways to ensure the user is not nil or that the user has to be authorized to perform this action. 
This of course will work but it's not the cleanest solution.

Also, you could use `.nil?` and you don't need to find the user when you have `current_user` available in this context - it already has found the user for you.

gahermar replied:
I must say i need to look more on cancancan ....

i was repair nil and remove finding user.

jstrojil commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,19 @@
+    if current_user.is_a?(Mentor) && current_user.kids.empty?
+      # if a menotr has no kids yet assigned, go to available kids
+      redirect_to available_kids_path
+    elsif current_user.is_a?(Teacher) && current_user.mentor_matchings.pluck(:state).include?('pending')
```
This is a way but the performance could be better-using database query -> using where or exists?

jstrojil replied:
Ah I meant something in this manner (using Activerecord) `current_user.mentor_matchings.exists?(state: 'pending')`

gahermar replied:
Ah this. ok. I changed it. And now question - we have same things in application controller after login. Change it too? I just ask because it is not my code but on the other side it is improve.

jstrojil commented on app/views/layouts/application.html.haml:
```diff
@@ -46,39 +47,47 @@
 
-
-    #sidebar-wrapper
+    - if (user_signed_in? && current_user.terms_of_use_accepted == false)
```
`(` not needed. Also please use negation operator instead of comparing to false

jstrojil commented on app/views/layouts/application.html.haml:
```diff
@@ -46,39 +47,47 @@
-            .contextual_links_panel.panel.panel-default.hidden-xs.hidden-sm
-              = render 'layouts/actions'
-            = yield :sidebar
+    -else
```
- else

jstrojil commented on spec/requests/site_configuration_spec.rb:
```diff
@@ -25,4 +25,32 @@
     expect(page).to have_css('h1', text: 'last1, first1')
     expect(page).to have_css('h2', text: 'Gesprächsdokumentationen')
   end
+
```
It would be cool to check if
- After updating the terms, the `terms_of_use_accepted` gets reset for other users that have previously accepted the terms
- Checking whether the user's `terms_of_use_accepted` change from false to true on the database level

gahermar replied:
I think the first check what you want is test on line 29 - i create admin and teacher (both with akcepted = true) then i login with admin - did some change in terms of use and save. Logout and login with teacher and check if there is showed page with term of use.

Second check - i tried to do new test in same file on line 57 - similar function but at the end i was reload teacher and check terms_of_use_accepted. Did you mean that?

jstrojil commented on config/locales/future_kids.de.yml:
```diff
@@ -247,6 +247,9 @@ de:
         title: Titel des Browsertabs
         css: Seitenspezifisches CSS
+        terms_of_use_accept:
+          terms_of_use_headline: "Bitte akzeptieren sie die geändertern Nutzungsbedingungen"
```
Just a nit, `"` are not needed

MartinGaher committed:
Pull request comments 

repairs

panterch commented:
Dear @MartinGaher 

Thanks for your work - very much appreciated!

I've checked out the branch and did some acceptance testing. Here are my findings:

The following acceptance criteria seems not to be fulfilled:

_Werden die Nutzungsbedingungen erneut geändert, müssen diese erneut angezeigt und akzeptiert werden_

In the implementation notes there were some hints how to achive this. Feel free to find another way of implementing it, that fulfills the same criterias:

_Auf dem site model soll aufgezeichnet, wann die Nutzungsbedingungen das letzte mal geändert wurden. Dies kann z.B. über den "Dirty" state des Feldes gemacht werden: https://api.rubyonrails.org/classes/ActiveRecord/AttributeMethods/Dirty.html

Auf dem User Model muss ebenfalls aufgezeichnet werden, wann Benutzer*innen das letzte Mal die Nutzungsbedingungen akzeptiert haben._

I tested by logging in as admin, accepting the tos, then changing the tos, logging out and login in again. It should have displayed me the tos again but did not.

MartinGaher committed:
Repair - after change tos enforce accept for all users

MartinGaher committed:
add footer to page with tos

jstrojil commented on app/views/layouts/application.html.haml:
```diff
@@ -46,39 +47,55 @@
 
-
-    #sidebar-wrapper
+    - if user_signed_in? && current_user.terms_of_use_accepted != true
```
`!current_user.terms_of_use_accepted`

gahermar replied:
done

jstrojil commented on app/controllers/sites_controller.rb:
```diff
@@ -18,7 +18,13 @@ def show
+      @site.terms_of_use_changed_date = DateTime.now
+    end
+
+    if @site.save
```
I see one little issue here - What if theoretically speaking saving failed. In this case, we would still set `terms_of_use_accepted` to `false` for all users. 
This is not a big issue but it could be maybe addressed as well.

panterch replied:
There's a convention to end date fields with `_at`. Maybe you rename this to `site.terms_of_use_changed_at` to make the code more readable.

gahermar replied:
Ok. when i changed logic to dates it is unnecessarily now.

i changed column names to _at

panterch commented on db/migrate/20220323125356_add_terms_of_use_accepted_date_to_users.rb:
```diff
@@ -0,0 +1,5 @@
+class AddTermsOfUseAcceptedDateToUsers < ActiveRecord::Migration[6.1]
+  def change
+    add_column :users, :terms_of_use_accepted_date, :datetime
```
There's a convention to end date fields with _at. Maybe you rename this to terms_of_use_accepted_at to make the code more readable.

gahermar replied:
i changed column names to _at

MartinGaher committed:
review repairs

Change names tos columns in Db, compute accepted from dates + some small changes

MartinGaher committed:
DB defaults on tos dates

gahermar commented:
Hi,
i have question about default values. Now i changed columns in DB to have default values to datetime.now. It is ok or it is better to let it nil and do some check in code? It depends on how is user created. If there is some acceptation during creation or not or if some other user create user account?
